File: ios/OakApp/Features/Artifact/PokemonUsagePane.swift
import SwiftUI

/// Full species usage drill-in for a Pokémon artifact (Summary's sibling tab).
/// Same lists as the Usage tab: ladder, snapshot, six share rows, Apply.
struct PokemonUsagePane: View {
  let slug: String
  var onOpen: (EntityKind, String) -> Void

  @Environment(\.services) private var services
  @State private var model: PokemonUsageModel?

  var body: some View {
    VStack(alignment: .leading, spacing: Theme.Spacing.lg) {
      Picker("Ladder", selection: ladderBinding) {
        ForEach(UsageLadder.allCases, id: \.self) { ladder in
          Text(ladder.title).tag(ladder)
        }
      }
      .pickerStyle(.segmented)
      .accessibilityLabel("Usage ladder")

      if let model {
        if model.isLoading && model.detail == nil {
          ProgressView()
            .frame(maxWidth: .infinity)
            .accessibilityLabel("Loading")
        } else {
          UsageSpeciesStack(
            detail: model.detail ?? .unavailable,
            slug: slug,
            onOpen: onOpen
          )
        }
      }
    }
    .frame(maxWidth: .infinity, alignment: .leading)
    .task(id: slug) {
      let next = PokemonUsageModel(slug: slug, usage: services.usage)
      model = next
      await next.load()
    }
  }

  private var ladderBinding: Binding<UsageLadder> {
    Binding(
      get: { model?.ladder ?? .doubles },
      set: { next in
        Task { await model?.selectLadder(next) }
      }
    )
  }
}

/// Snapshot + share lists + Apply, laid out for a parent `ScrollView`
/// (the artifact sheet already scrolls — do not nest a `List`).
struct UsageSpeciesStack: View {
  let detail: UsageSpeciesResponse
  let slug: String
  var onOpen: (EntityKind, String) -> Void

  var body: some View {
    if detail.available == false {
      VStack(alignment: .leading, spacing: Theme.Spacing.sm) {
        Text("Usage unavailable")
          .font(Theme.display(.title3))
          .foregroundStyle(Theme.textPrimary)
        Text(UsageViewModel.unavailableCopy)
          .font(Theme.body(.footnote))
          .foregroundStyle(Theme.textSecondary)
      }
    } else if detail.found != true {
      VStack(alignment: .leading, spacing: Theme.Spacing.sm) {
        Text("No usage set")
          .font(Theme.display(.title3))
          .foregroundStyle(Theme.textPrimary)
        Text("No Champions usage set is listed for this species. It may not be on the Champions roster.")
          .font(Theme.body(.footnote))
          .foregroundStyle(Theme.textSecondary)
      }
    } else {
      foundBody
    }
  }

  @ViewBuilder
  private var foundBody: some View {
    VStack(alignment: .leading, spacing: Theme.Spacing.xs) {
      if let season = detail.season {
        LabeledContent("Season", value: season)
      }
      if let fetchedAt = detail.fetchedAt {
        LabeledContent("Updated", value: formatUsageFetchedAt(fetchedAt))
      }
      if let attribution = detail.attribution, !attribution.isEmpty {
        let parts = parseUsageAttribution(attribution)
        LabeledContent("Source", value: parts.source)
        if let legal = parts.legal {
          Text(legal)
            .font(Theme.body(.caption))
            .foregroundStyle(Theme.textMuted)
        }
      }
    }

    usageGroup("Moves", kind: .moves, entries: detail.moves)
    usageGroup("Items", kind: .items, entries: detail.items)
    usageGroup("Abilities", kind: .abilities, entries: detail.abilities)
    usageGroup("Natures", kind: .natures, entries: detail.natures)
    usageGroup("Spreads", kind: .spreads, entries: detail.spreads)
    usageGroup("Teammates", kind: .teammates, entries: detail.teammates)

    ApplyChampionsSetButton(species: detail.slug ?? slug)
  }

  @ViewBuilder
  private func usageGroup(_ title: String, kind: UsageListKind, entries: [UsageEntry]?) -> some View {
    if let entries, !entries.isEmpty {
      VStack(alignment: .leading, spacing: Theme.Spacing.sm) {
        Text(title)
          .instrumentLabel()
          .foregroundStyle(Theme.textMuted)
        ForEach(Array(entries.enumerated()), id: \.offset) { _, entry in
          if let route = UsageDexLink.route(kind: kind, name: entry.name) {
            Button {
              onOpen(route.kind, route.query)
            } label: {
              UsageShareRow(name: entry.name, pct: entry.pct, showsChevron: true)
            }
            .buttonStyle(.plain)
            .accessibilityHint("Opens \(entry.name)")
          } else {
            UsageShareRow(name: entry.name, pct: entry.pct, showsChevron: false)
          }
        }
      }
    }
  }
}

struct UsageShareRow: View {
  let name: String
  let pct: Double?
  var showsChevron: Bool

  var body: some View {
    HStack {
      Text(name)
        .foregroundStyle(Theme.textPrimary)
      Spacer()
      if let pct {
        Text(String(format: "%.1f%%", pct))
          .font(Theme.mono(.footnote))
          .foregroundStyle(Theme.textSecondary)
      }
      if showsChevron {
        Image(systemName: "chevron.right")
          .font(.caption.weight(.semibold))
          .foregroundStyle(Theme.textMuted)
      }
    }
    .contentShape(Rectangle())
  }
}

File: android/app/src/main/kotlin/ai/gowtam/oak/features/artifact/PokemonUsagePane.kt
package ai.gowtam.oak.features.artifact

import ai.gowtam.oak.app.LocalServices
import ai.gowtam.oak.features.usage.UsageDexLink
import ai.gowtam.oak.features.usage.UsageListKind
import ai.gowtam.oak.features.usage.formatUsageFetchedAt
import ai.gowtam.oak.features.usage.parseUsageAttribution
import ai.gowtam.oak.ui.LocalOakColors
import ai.gowtam.oak.ui.OakButton
import ai.gowtam.oak.ui.OakButtonStyle
import ai.gowtam.oak.ui.OakSpacing
import ai.gowtam.oak.wire.EntityKind
import ai.gowtam.oak.wire.TeamMember
import ai.gowtam.oak.wire.UsageEntry
import ai.gowtam.oak.wire.UsageLadder
import ai.gowtam.oak.wire.UsageSpecies
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.FilterChipDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch

private const val UNAVAILABLE_COPY =
    "Live Champions usage is unavailable right now. Chat, Dex, Teams, and Calc still work — try this page again in a bit."

/**
 * Full species usage drill-in for a Pokémon artifact (Summary's sibling tab).
 */
@Composable
fun PokemonUsagePane(
    slug: String,
    onOpen: (EntityKind, String) -> Unit,
    modifier: Modifier = Modifier,
    onAddToTeam: ((TeamMember) -> Unit)? = null,
    onApplySpecies: ((String) -> Unit)? = null,
) {
    val usage = LocalServices.current?.usage
    val oak = LocalOakColors.current
    val loader = remember(usage) { usage?.let { PokemonUsageLoader(it) } }
    val idle = remember { kotlinx.coroutines.flow.MutableStateFlow(PokemonUsageLoader.State()) }
    val state by (loader?.state ?: idle).collectAsState()
    var note by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()

    LaunchedEffect(slug, loader) {
        loader?.load(slug)
    }

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(OakSpacing.sm),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(OakSpacing.xs)) {
            FilterChip(
                selected = state.ladder == UsageLadder.Doubles,
                onClick = { scope.launch { loader?.load(slug, UsageLadder.Doubles) } },
                label = { Text("Doubles") },
                colors = FilterChipDefaults.filterChipColors(
                    selectedContainerColor = oak.accentSoft,
                    selectedLabelColor = oak.accent,
                ),
            )
            FilterChip(
                selected = state.ladder == UsageLadder.Singles,
                onClick = { scope.launch { loader?.load(slug, UsageLadder.Singles) } },
                label = { Text("Singles") },
                colors = FilterChipDefaults.filterChipColors(
                    selectedContainerColor = oak.accentSoft,
                    selectedLabelColor = oak.accent,
                ),
            )
        }

        val detail = state.detail
        when {
            loader == null -> {
                Text(UNAVAILABLE_COPY, color = oak.textMuted)
            }
            state.loading && detail == null -> {
                CircularProgressIndicator(color = oak.accent)
            }
            detail == null || !detail.available -> {
                Text("Usage unavailable", style = MaterialTheme.typography.titleMedium, color = oak.textStrong)
                Text(detail?.error ?: UNAVAILABLE_COPY, color = oak.textMuted)
            }
            detail.found != true -> {
                Text("No usage set", style = MaterialTheme.typography.titleMedium, color = oak.textStrong)
                Text(
                    "No Champions usage set is listed for this species. It may not be on the Champions roster.",
                    color = oak.textMuted,
                )
            }
            else -> {
                ArtifactUsageFound(
                    detail = detail,
                    slug = slug,
                    onOpen = onOpen,
                    onApply = {
                        if (onApplySpecies != null) {
                            onApplySpecies(it)
                        } else {
                            note = "Sign in to apply this Champions set to a team."
                        }
                    },
                )
            }
        }
        note?.let { Text(it, color = oak.textMuted, style = MaterialTheme.typography.bodySmall) }
    }
}

@Composable
private fun ArtifactUsageFound(
    detail: UsageSpecies,
    slug: String,
    onOpen: (EntityKind, String) -> Unit,
    onApply: (String) -> Unit,
) {
    val oak = LocalOakColors.current
    val title = detail.savedName ?: detail.slug ?: slug
    Text(title, style = MaterialTheme.typography.titleMedium.copy(fontWeight = FontWeight.SemiBold), color = oak.textStrong)
    if (detail.season != null || detail.fetchedAt != null) {
        UsageSnapshotLine(season = detail.season, fetchedAt = detail.fetchedAt)
    }
    detail.attribution?.takeIf { it.isNotBlank() }?.let { UsageSourceLine(it) }
    UsageShareSection("Moves", detail.moves, UsageListKind.MOVES, onOpen)
    UsageShareSection("Items", detail.items, UsageListKind.ITEMS, onOpen)
    UsageShareSection("Abilities", detail.abilities, UsageListKind.ABILITIES, onOpen)
    UsageShareSection("Natures", detail.natures, UsageListKind.NATURES, onOpen)
    UsageShareSection("Spreads", detail.spreads, UsageListKind.SPREADS, onOpen)
    UsageShareSection("Teammates", detail.teammates, UsageListKind.TEAMMATES, onOpen)
    val applySlug = detail.slug?.takeIf { it.isNotBlank() } ?: slug
    OakButton(
        onClick = { onApply(applySlug) },
        style = OakButtonStyle.Primary,
        modifier = Modifier.fillMaxWidth(),
    ) {
        Text("Apply this Champions set")
    }
}

@Composable
private fun UsageShareSection(
    title: String,
    entries: List<UsageEntry>,
    kind: UsageListKind,
    onOpen: (EntityKind, String) -> Unit,
) {
    if (entries.isEmpty()) return
    val oak = LocalOakColors.current
    Text(title, style = MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.SemiBold), color = oak.textStrong)
    for (entry in entries) {
        val pct = entry.pct?.let { "${"%.1f".format(it)}%" } ?: "—"
        val target = UsageDexLink.route(kind, entry.name)
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .then(
                    if (target != null) {
                        Modifier.clickable(onClick = { onOpen(target.kind, target.query) }, onClickLabel = "Open ${entry.name}")
                    } else {
                        Modifier
                    },
                )
                .padding(vertical = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(OakSpacing.sm),
        ) {
            Text(entry.name, style = MaterialTheme.typography.bodyMedium, color = oak.text, modifier = Modifier.weight(1f))
            Text(pct, style = MaterialTheme.typography.bodyMedium, color = oak.textMuted)
            if (target != null) {
                Icon(Icons.Filled.ChevronRight, contentDescription = null, tint = oak.textMuted)
            }
        }
    }
}

@Composable
private fun UsageSnapshotLine(season: String?, fetchedAt: Long?) {
    val oak = LocalOakColors.current
    val liveLine = listOfNotNull("Live", season?.takeIf { it.isNotBlank() }).joinToString(" · ")
    val fetchedLabel = fetchedAt?.let { formatUsageFetchedAt(it) }
    Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
        if (liveLine.isNotEmpty()) {
            Text(liveLine.uppercase(), style = MaterialTheme.typography.labelSmall, color = oak.textMuted)
        }
        fetchedLabel?.let { Text(it, style = MaterialTheme.typography.bodySmall, color = oak.textMuted) }
    }
}

@Composable
private fun UsageSourceLine(attribution: String) {
    val oak = LocalOakColors.current
    val parts = parseUsageAttribution(attribution)
    Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
        Text("SOURCE", style = MaterialTheme.typography.labelSmall, color = oak.textMuted)
        Text(parts.source, style = MaterialTheme.typography.bodySmall, color = oak.text)
        parts.legal?.let { Text(it, style = MaterialTheme.typography.bodySmall, color = oak.textMuted) }
    }
}
